speechbrain/utils/logger.py

```diff
@@ -62,24 +64,105 @@
+    and provides additional functionality for controlling logging in multi-process
+    environments, with the option to limit logs to the main process only.
 
-    Writes and prints should be passed through tqdm.tqdm.write
-    so that the tqdm progressbar doesn't get messed up.
+    This class is heavily inspired by HuggingFace Accelerate toolkit:
+    https://github.com/huggingface/accelerate/blob/85b1a03552cf8d58e036634e004220c189bfb247/src/accelerate/logging.py#L22
     """
 
-    def emit(self, record):
-        """TQDM compatible StreamHandler."""
-        try:
-            msg = self.format(record)
-            stream = self.stream
-            tqdm.tqdm.write(msg, end=self.terminator, file=stream)
-            self.flush()
-        except RecursionError:
-            raise
-        except Exception:
-            self.handleError(record)
+    @staticmethod
+    def _should_log(main_process_only: bool) -> bool:
+        """
+        Determines if logging should occur based on whether the code is running
+        on the main process or not.
+
+        Parameters
+        ----------
+        main_process_only : bool
+            A flag indicating if logging should be restricted to the main process.
+
+        Returns
+        -------
+        bool
+            True if logging should be performed (based on the process and the flag),
+            False otherwise.
+        """
+        return not main_process_only or (
+            main_process_only and if_main_process()
+        )
+
+    def log(self, level: int, msg: str, *args: tuple, **kwargs: dict):
+        """
+        Logs a message with the specified log level, respecting the `main_process_only`
+        flag to decide whether to log based on the current process.
+
+        Parameters
+        ----------
+        level : int
+            Logging level (e.g., logging.INFO, logging.WARNING).
+        msg : str
+            The message to log.
+        *args : tuple
+            Additional positional arguments passed to the logger.
+        **kwargs : dict
+            Additional keyword arguments passed to the logger, including:
+            - main_process_only (bool): If True, log only from the main process (default: True).
+            - stacklevel (int): The stack level to use when logging (default: 2).
+
+        Notes
+        -----
+        If `main_process_only` is True, the log will only be written if the current process
+        is the main process, as determined by `if_main_process()`.
+        """
+        main_process_only = kwargs.pop("main_process_only", True)
+        kwargs.setdefault("stacklevel", 2)
+
+        if self.isEnabledFor(level):
+            if self._should_log(main_process_only):
+                msg, kwargs = self.process(msg, kwargs)
+                self.logger.log(level, msg, *args, **kwargs)
+
+    @functools.lru_cache(None)
+    def warning_once(self, *args: tuple, **kwargs: dict):
```

Comment:
Neat trick, are there any places in the code where we should replace the warning with this one?

Reply:
Yes --- I added one call of this function in the repo. :)